File: python/tests/sorting_benchmark.py
import random, time, json, tracemalloc, gc

class SortingBenchmarkTest:
    name = "SortingBenchmark"

    def __init__(self, n=10000, ops_per_iter=None):
        self.n = n
        self.ops_per_iter = ops_per_iter or n  # for consistency with other tests

    def workload(self):
        data = [random.randint(0, 1_000_000) for _ in range(self.n)]
        data.sort()
        return sum(data[:100])  # checksum for sanity

    def run_once(self):
        gc.collect()
        tracemalloc.start()
        t0 = time.perf_counter()

        checksum = self.workload()

        t1 = time.perf_counter()
        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        duration = t1 - t0
        # Estimate operations as n * log2(n)
        ops_done = int(self.n * (self.n.bit_length()))

        return {
            "time_s": duration,
            "ops": ops_done,
            "ops_per_sec": ops_done / duration if duration > 0 else float("inf"),
            "mem_peak_bytes": peak,
            "checksum": checksum
        }

    def run(self, runs=5):
        results = [self.run_once() for _ in range(runs)]
        times = sorted(r["time_s"] for r in results)
        ops = sorted(r["ops_per_sec"] for r in results)
        peaks = sorted(r["mem_peak_bytes"] for r in results)
        mid = len(times) // 2

        return {
            "name": self.name,
            "runs": runs,
            "median_time_s": times[mid],
            "median_ops_per_sec": ops[mid],
            "median_peak_mem_bytes": peaks[mid],
            "raw": results
        }

if __name__ == "__main__":
    result = SortingBenchmarkTest(10_000).run()
    with open("results_python_sorting_benchmark.json", "w") as f:
        json.dump(result, f, indent=2)

File: ruby/tests/sorting_benchmark.rb
require "json"
require "time"

class SortingBenchmarkTest
  attr_reader :name

  def initialize(n = 10000)
    @name = "SortingBenchmark"
    @n = n
  end

  def run_once
    data = Array.new(@n) { rand(1_000_000) }
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    data.sort!
    duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0
    ops = @n * Math.log2(@n)
    {
      "time_s" => duration,
      "ops" => ops,
      "ops_per_sec" => ops / (duration > 0 ? duration : 1e-9),
      "checksum" => data[0, 100].sum
    }
  end

  def run(runs = 5, intrations = nil)
    results = Array.new(runs) { run_once }
    times = results.map { |r| r["time_s"] }.sort
    ops = results.map { |r| r["ops_per_sec"] }.sort
    median = runs / 2
    {
      "name" => @name,
      "runs" => runs,
      "median_time_sec" => times[median],
      "median_ops_per_sec" => ops[median],
      "raw" => results
    }
  end
end

if __FILE__ == $0
  test = SortingBenchmarkTest.new(10000)
  result = test.run
  File.write("results_ruby_sorting_benchmark.json", JSON.pretty_generate(result))
end
